### DISCUSSION TITLE
An efficient modeling interface in Python with support for HIGHS

### DISCUSSION THREAD

**Discussion by metab0t** [ERGO-Code/HiGHS#1708] [Show and tell]
Today, I want to share a newly developed modeling interface for Python: https://github.com/metab0t/PyOptInterface with support for HiGHS.

It can be installed with [our PyPI package of HiGHS](https://github.com/metab0t/highsbox) to use HiGHS in Python out of the box:
```
pip install pyoptinterface[highs]
```

It provides high-level modeling capability by using the C API of HiGHS directly and achieves competitive performance to construct large scale models.

It has built-in support for [N-dimensional containers](https://metab0t.github.io/PyOptInterface/container.html) for variables and constraints. The variables and constraints can also be used with Numpy to formulate multi-dimensional array.

This is an introductory [example](https://metab0t.github.io/PyOptInterface/numpy.html) to use PyOptInterface and HiGHS to solve N-queens problem:
```python
import numpy as np
import pyoptinterface as poi
from pyoptinterface import highs

model = highs.Model()

N = 8

x = np.empty((N, N), dtype=object)
for i in range(N):
    for j in range(N):
        x[i, j] = model.add_variable(domain=poi.VariableDomain.Binary)

for i in range(N):
    # Row and column
    model.add_linear_constraint(poi.quicksum(x[i, :]), poi.Eq, 1.0)
    model.add_linear_constraint(poi.quicksum(x[:, i]), poi.Eq, 1.0)
for i in range(-N+1, N):
    # Diagonal
    model.add_linear_constraint(poi.quicksum(x.diagonal(i)), poi.Leq, 1.0)
    # Anti-diagonal
    model.add_linear_constraint(poi.quicksum(np.fliplr(x).diagonal(i)), poi.Leq, 1.0)

model.optimize()

print("Termination status:", model.get_model_attribute(poi.ModelAttribute.TerminationStatus))

get_v = np.vectorize(lambda x: model.get_value(x))
x_value = get_v(x)

print(x_value.astype(int))
```

More detailed documentation is [available online](https://metab0t.github.io/PyOptInterface/).

Thanks for @jajhall for developing such an amazing open source optimizer!

Feedbacks and stars are all welcome 😄.

**Comment by jajhall** [comment]
Thanks for developing this Python packaging of HiGHS. I didn't find any documentation for the methods in the API. 

Note that HiGHS would not have been developed without @galabovaa, or the writers of its solvers: Qi Huangfu, @lschork2, @lgottwald and @feldmeier.

**Reply by metab0t** [reply]
Yep, the autogenerated API documentation is still sparse.

The main APIs for modeling interface are documented on [this page](https://metab0t.github.io/PyOptInterface/common_model_interface.html)

The Python packaging of HiGHS enables users to have instant access to HiGHS without downloading binaries manually.

**Reply by jajhall** [reply]
OK, thanks. I understand why you're focusing on the top level modelling interface
